larq/optimizers_v2.py

```diff
@@ -118,32 +94,34 @@ def _get_optimizer(self, var):
         if optimizer_index == self.DEFAULT_OPT_INDEX:
             return self.default
         else:
-            return self.pred_opt_pairs[optimizer_index][1]  # from (pred, opt) tuple
+            return self.pred_opt_pairs[optimizer_index][1]  # [1] from (pred, opt) tuple
 
     def _compute_var_opt_mapping(self, grads_and_vars):
         """Compute a unique mapping from variables to optimizers.
         
-        Also yield a new iterator for the original `grads_and_vars`, which can then be
-        used by `apply_gradients()`.
+        Also yield a new iterator for the original `grads_and_vars` so that it is not
+        depleted once `apply_gradients()` needs it.
         """
 
         self.var_opt_mapping = {}
 
         for grad, var in grads_and_vars:
-            num_opts = 0
+            num_optimizers = 0
 
+            # Find the optimizer(s) that want to claim this variable
             for optimizer_index, (predicate, _) in enumerate(self.pred_opt_pairs):
                 if predicate(var):
                     self.var_opt_mapping[var.name] = optimizer_index
-                    num_opts += 1
+                    num_optimizers += 1
 
-            if num_opts > 1:
+            if num_optimizers > 1:
                 raise ValueError(f"Variable `{var}` claimed by multiple optimizers.")
-            if num_opts == 0:
+            if num_optimizers == 0:
                 self.var_opt_mapping[var.name] = self.DEFAULT_OPT_INDEX
                 if self.default is None:
                     warnings.warn(f"No `default` provided to train variable `{var}`.")
 
+            # Yielding a new grads_and_vars iterator for use by `apply_gradients()`
             yield ((grad, var))
```

Comment:
This makes it somewhat confusing to grasp the execution path. We can also cast the iterator to a list (see https://github.com/tensorflow/tensorflow/issues/33345 for context).

Reply:
Changed in 7aa8624.